In [1]:
import pandas as pd
import os


In [2]:
lista_cpfs = os.listdir("2019_cpf")

lista_cpfs

['CENTRO_OESTE2019ID.dta.pkl',
 'MG_ES_RJ2019ID.dta.pkl',
 'NORDESTE2019ID.dta.pkl',
 'NORTE2019ID.dta.pkl',
 'SP2019ID.dta.pkl',
 'SUL2019ID.dta.pkl']

In [3]:
cpf_total = pd.DataFrame()
for cpf_file in lista_cpfs:
    df_cpf = pd.read_pickle("2019_cpf\\"+cpf_file)
    cpf_total = pd.concat([cpf_total,df_cpf])

In [4]:
cpf_total

CPF     identificad radiccnpj dtadmissao  diadesli mesdesli  \
0         04986073133  10853605000189  10853605   01042018       7.0      6.0   
1         52549070453  10759501000290  10759501   25092018      18.0      2.0   
2         05298743188  11117726000125  11117726   19112018      21.0      6.0   
3         87176130191  07338504000165  07338504   10012019      15.0      5.0   
4         46514004857  07338504000165  07338504   06082018       3.0      4.0   
...               ...             ...       ...        ...       ...      ...   
12284025  01429454040  91481432000117  91481432   13112017       NaN      NaN   
12284026  01438629010  94088952000152  94088952   03122018       NaN      NaN   
12284027  01221867024  92695691000103  92695691   06082018       NaN      NaN   
12284028  01843760096  92695691000103  92695691   15072019       NaN      NaN   
12284029  36580990899  61099834071292  61099834   01052019       NaN      NaN   

           remmedr ocup2002  
0          1507.05   848520  
1          1046.00   632615  
2          1094.63   632015  
3          2093.00   782510  
4          1740.59   783225  
...            ...      ...  
12284025  10684.89   142305  
12284026   2113.30   311105  
12284027   3546.55   411010  
12284028   3272.28   411010  
12284029   1661.27   521115  

[66667417 rows x 8 columns]

In [4]:
csup = pd.read_stata("D:\\datasets\\coorte13_ces_enem.dta")

In [6]:
#len(pd.unique(csup.no_curso))

In [ ]:
#co_curso_list = [18722, 2899, 63414, 19532, 2011, 11220, 9633, 5691, 7939, 5717]
#co_curso_list = [18722, 2899, 63414, 19532, 2011, 11220, 9633, 5691, 7939]
#csup = csup[csup['co_curso'].isin(co_curso_list)]
#pd.unique(csup.co_ies)

array([  55,  322, 1161,  546,   22,  461,  374, 2129, 2183], dtype=int16)

55 USP; 322 UNIP; 1161 INSPER; 546 PUC SP; 213 FECAP; 22 MACK; 461 FAAP; 374 FMU;  2129 EESP; 2183 fundacao santo andre

In [8]:
csup.drop_duplicates(subset=["nu_cpf"]).columns

Index(['co_aluno', 'nu_cpf', 'sexo', 'cor_raca', 'cor_raca_ces',
       'cor_raca_enem', 'conclusao_em', 'ano_conclusao_em', 'idade',
       'ano_nascimento', 'in_enem', 'nota_cn', 'nota_ch', 'nota_lc', 'nota_mt',
       'notas', 'nota_redacao', 'edupar', 'escpai', 'escmae', 'rdpc',
       'log_rdpc', 'renda', 'qt_hab', 'imovel', 'trabalha', 'escola_ef',
       'escola_em', 'co_curso', 'vinculo_2013', 'vinculo_2014', 'vinculo_2015',
       'vinculo_2016', 'vinculo_2017', 'vinculo_2018', 'vinculo_2019',
       'co_ies', 'categoria_adm', 'organizacao_acad', 'grau', 'modalidade',
       'co_cine_rotulo', 'no_cine_rotulo', 'co_cine_area_detalhada',
       'no_cine_area_detalhada', 'co_cine_area_especifica',
       'no_cine_area_especifica', 'co_cine_area_geral', 'no_cine_area_geral',
       'reserva_vagas_2013', 'turno_2013', 'turno_2014', 'turno_2015',
       'turno_2016', 'turno_2017', 'turno_2018', 'turno_2019',
       'financiamento_2013', 'financiamento_2014', 'financiamento_2015',
  

In [5]:
cpf_total["nu_cpf"]=cpf_total["CPF"].astype(float) 
cpf_total

CPF     identificad radiccnpj dtadmissao  diadesli mesdesli  \
0         04986073133  10853605000189  10853605   01042018       7.0      6.0   
1         52549070453  10759501000290  10759501   25092018      18.0      2.0   
2         05298743188  11117726000125  11117726   19112018      21.0      6.0   
3         87176130191  07338504000165  07338504   10012019      15.0      5.0   
4         46514004857  07338504000165  07338504   06082018       3.0      4.0   
...               ...             ...       ...        ...       ...      ...   
12284025  01429454040  91481432000117  91481432   13112017       NaN      NaN   
12284026  01438629010  94088952000152  94088952   03122018       NaN      NaN   
12284027  01221867024  92695691000103  92695691   06082018       NaN      NaN   
12284028  01843760096  92695691000103  92695691   15072019       NaN      NaN   
12284029  36580990899  61099834071292  61099834   01052019       NaN      NaN   

           remmedr ocup2002        nu_cpf  
0          1507.05   848520  4.986073e+09  
1          1046.00   632615  5.254907e+10  
2          1094.63   632015  5.298743e+09  
3          2093.00   782510  8.717613e+10  
4          1740.59   783225  4.651400e+10  
...            ...      ...           ...  
12284025  10684.89   142305  1.429454e+09  
12284026   2113.30   311105  1.438629e+09  
12284027   3546.55   411010  1.221867e+09  
12284028   3272.28   411010  1.843760e+09  
12284029   1661.27   521115  3.658099e+10  

[66667417 rows x 9 columns]

In [6]:
#merged_csup = csup[["nu_cpf","co_curso"]].merge(cpf_total[["radiccnpj","dtadmissao","diadesli","mesdesli","nu_cpf"]].drop_duplicates(subset=["nu_cpf"]),on="nu_cpf",how="inner")
#merged_csup = csup[["nu_cpf","co_curso"]].merge(cpf_total[["radiccnpj","dtadmissao","diadesli","mesdesli","nu_cpf","remmedr"]],on="nu_cpf",how="inner")
merged_csup = csup.merge(cpf_total[["radiccnpj","dtadmissao","diadesli","mesdesli","nu_cpf","remmedr","ocup2002"]],on="nu_cpf",how="inner")

In [7]:
merged_csup.to_pickle("D:\\datasets\\merged_csuprais.pkl")

In [8]:
merged_csup#.merge(merged_csup,how="cross")

co_aluno        nu_cpf       sexo cor_raca cor_raca_ces  \
0        4.000113e+11  2.748279e+09   Feminino   Branca       Branca   
1        4.000113e+11  2.748279e+09   Feminino   Branca       Branca   
2        4.000156e+11  1.184102e+10   Feminino   Branca       Branca   
3        4.000158e+11  3.999433e+10  Masculino   Branca          NaN   
4        4.000158e+11  1.404696e+10  Masculino   Branca       Branca   
...               ...           ...        ...      ...          ...   
2187096  4.000118e+11  4.006206e+10  Masculino   Branca       Branca   
2187097  4.000158e+11  9.643702e+09   Feminino   Branca       Branca   
2187098  4.000151e+11  3.914525e+10  Masculino   Branca       Branca   
2187099  4.000102e+11  4.189562e+10  Masculino   Branca       Branca   
2187100  4.000102e+11  4.189562e+10  Masculino   Branca       Branca   

        cor_raca_enem             conclusao_em  ano_conclusao_em  idade  \
0              Branca            Ja conclui EM            2009.0   21.0   
1              Branca            Ja conclui EM            2009.0   21.0   
2              Branca    Concluinte EM em 2012               NaN   18.0   
3              Branca    Concluinte EM em 2012               NaN   19.0   
4              Branca  Concluinte EM apos 2012               NaN   19.0   
...               ...                      ...               ...    ...   
2187096           NaN                      NaN               NaN   22.0   
2187097        Branca            Ja conclui EM            2011.0   21.0   
2187098        Branca            Ja conclui EM            2011.0   18.0   
2187099           NaN                      NaN               NaN   21.0   
2187100           NaN                      NaN               NaN   21.0   

         ano_nascimento  ... apoio_social_2016  apoio_social_2017  \
0                  1992  ...               Nao                Nao   
1                  1992  ...               Nao                Nao   
2                  1995  ...               Nao                NaN   
3                  1994  ...               NaN                NaN   
4                  1994  ...               Sim                Sim   
...                 ...  ...               ...                ...   
2187096            1991  ...               Nao                Nao   
2187097            1992  ...               Nao                Nao   
2187098            1995  ...               Nao                NaN   
2187099            1992  ...               Nao                NaN   
2187100            1992  ...               Nao                NaN   

         apoio_social_2018  apoio_social_2019  radiccnpj  dtadmissao  \
0                      NaN                NaN   26192813    11022019   
1                      NaN                NaN   32483651    13052019   
2                      NaN                NaN   06137579    01062017   
3                      NaN                NaN   59527788    10092018   
4                      Nao                Sim   22896431    04112019   
...                    ...                ...        ...         ...   
2187096                NaN                NaN   06252818    18032013   
2187097                NaN                NaN   26748019    12122016   
2187098                NaN                NaN   55495444    09042018   
2187099                NaN                NaN   50719061    01102012   
2187100                NaN                NaN   30200247    06062019   

         diadesli mesdesli  remmedr ocup2002  
0            11.0      5.0  2299.99   239405  
1             NaN      NaN  2338.33   239405  
2            23.0      8.0  1299.07   253115  
3             NaN      NaN  2145.67   252205  
4             NaN      NaN  1730.98   212320  
...           ...      ...      ...      ...  
2187096       NaN      NaN  3714.81   314410  
2187097       NaN      NaN  1684.48   252705  
2187098       NaN      NaN  2538.79   410105  
2187099       5.0      6.0  2607.87   212420  
2187100       NaN      NaN  2738.55 

In [10]:
merged_csup.head().columns#[["nu_cpf","CPF"]].dropna(subset='CPF')

Index(['nu_cpf', 'co_curso', 'radiccnpj', 'dtadmissao', 'diadesli',
       'mesdesli'],
      dtype='object')

## Create unique identifier

In [14]:
import hashlib
import pandas as pd


# Convert all to string and create a unique identifier
merged_csup['unique_id'] = merged_csup.apply(lambda x: hashlib.sha256(f"{str(x['co_curso'])}_{str(x['radiccnpj'])}".encode()).hexdigest(), axis=1)


In [18]:
import pandas as pd


# Convert 'dtadmissao' to datetime format
#merged_csup['dtadmissao'] = pd.to_datetime(merged_csup['dtadmissao'], format='%d%m%Y')

merged_csup['diadesli'] = pd.to_numeric(merged_csup['diadesli'], errors='coerce')
merged_csup['mesdesli'] = pd.to_numeric(merged_csup['mesdesli'], errors='coerce')

# Function to construct the separation date safely
def get_desligamento(row):
    if pd.notna(row['diadesli']) and pd.notna(row['mesdesli']):
        try:
            return pd.Timestamp(year=2019, month=int(row['mesdesli']), day=int(row['diadesli']))
        except ValueError:
            return pd.NaT  # Handles invalid day/month combinations
    else:
        return pd.Timestamp("2020-01-01")  # Default for NaN values

# Apply the function row-wise
merged_csup['desligamento'] = merged_csup.apply(get_desligamento, axis=1)

merged_csup.head()

nu_cpf  co_curso radiccnpj dtadmissao  diadesli  mesdesli  \
0  2.748279e+09     15852  26192813 2019-02-11      11.0       5.0   
1  1.184102e+10     33738  06137579 2017-06-01      23.0       8.0   
2  3.999433e+10     49464  59527788 2018-09-10       NaN       NaN   
3  1.404696e+10     12808  22896431 2019-11-04       NaN       NaN   
4  8.032529e+09    122906  00394502 1998-06-15       NaN       NaN   

                                           unique_id desligamento  
0  4424e369d38ca58544e9b1d1627d27fe5bc892cdb29cf2...   2019-05-11  
1  95b93caee0ff07ea6a3e07f069e3716bb0cf0fecb683f1...   2019-08-23  
2  93f4c56267ba4953b02b1c84acfe75e38e3f3c51be3f10...   2020-01-01  
3  0cd71dfa5086be33ebc1d00f04f8b57b49237c08b9b359...   2020-01-01  
4  4617ae76afbc5e7d3949b95fe6ce24c759bcff23f55b0a...   2020-01-01

In [11]:
len(pd.unique(merged_csup.nu_cpf))/len(pd.unique(csup.nu_cpf))

0.5958920172501401

In [22]:
def get_colleagues(row, df):
    same_hash_group = df[df["unique_id"] == row["unique_id"]]  # Get same unique_id group
    overlapping = same_hash_group[
        (same_hash_group["dtadmissao"] <= row["desligamento"]) & 
        (same_hash_group["desligamento"] >= row["dtadmissao"]) &  # Overlap condition
        (same_hash_group["nu_cpf"] != row["nu_cpf"])  # Exclude itself
    ]
    return list(overlapping["nu_cpf"])

# Apply function for each row
#merged_csup["colleagues"] = merged_csup.apply(lambda row: get_colleagues(row, merged_csup), axis=1)

In [23]:
from tqdm import tqdm
import pandas as pd

# Ensure tqdm works inside pandas apply
tqdm.pandas()

# Apply function with progress tracking
#merged_csup["colleagues"] = merged_csup.progress_apply(lambda row: get_colleagues(row, merged_csup), axis=1)
teste = merged_csup.progress_apply(lambda row: get_colleagues(row, merged_csup), axis=1)


  0%|          | 786/1740520 [00:57<35:34:59, 13.58it/s]


KeyboardInterrupt: 

In [30]:
merged_csup.unique_id.value_counts()

unique_id
91d3642312d184e682f1c42215790eead6bc6a3261d20fd77e1e8067766b5a63    760
33e9ec4b9da2563577db0499d880171f324bdf43f9e19d3a42f3fbcb78719985    656
27b0f255c160eead9247a6e76be915bf752069a03c3935387f033113b3e0e1c7    532
8e4e28be189ea764038aa098d42c60618ec97f324de730d781440d2a39ef4cbf    434
844b62e34f6ed427be9618c93a1794e213a87872d98c69060de574636a9735a0    355
                                                                   ... 
cd95984ed58ad2049ca73b79b00e367f0842f7baf762491196a7477e0bea0435      1
aac477b29e8a9af058658592d74b864156bf69790e5189a345928082903139e5      1
c1f2b5cb198d36d55e0a5ae6e715645fa61abc92cdfb16ac7f7d3f8179f3abf9      1
df81e4d1bcbafb4c3e73b4f2ab062aacecbbe4501cbd4e736f13ea033d508273      1
9b27412dedeea54831215da61d1bf93252b5c75282520a0cfee19b8e925d8671      1
Name: count, Length: 1370530, dtype: int64

In [29]:
merged_csup[merged_csup.unique_id=="33e9ec4b9da2563577db0499d880171f324bdf43f9e19d3a42f3fbcb78719985"]

nu_cpf  co_curso radiccnpj dtadmissao  diadesli  mesdesli  \
1233029  2.251767e+10     96655  46379400 2015-05-11       NaN       NaN   
1233231  1.571883e+10     96655  46379400 2018-02-02       NaN       NaN   
1233283  2.965865e+10     96655  46379400 2014-06-04       NaN       NaN   
1233963  2.830239e+10     96655  46379400 2004-08-23       NaN       NaN   
1234741  1.813044e+10     96655  46379400 2012-02-01       NaN       NaN   
...               ...       ...       ...        ...       ...       ...   
1732750  1.274843e+10     96655  46379400 2014-07-14       NaN       NaN   
1733509  2.741082e+10     96655  46379400 2018-01-19       NaN       NaN   
1734227  2.739366e+10     96655  46379400 2018-02-01       NaN       NaN   
1734365  2.188159e+10     96655  46379400 2007-02-12       NaN       NaN   
1734448  2.540627e+10     96655  46379400 2008-02-13       NaN       NaN   

                                                 unique_id desligamento  
1233029  33e9ec4b9da2563577db0499d880171f324bdf43f9e19d...   2020-01-01  
1233231  33e9ec4b9da2563577db0499d880171f324bdf43f9e19d...   2020-01-01  
1233283  33e9ec4b9da2563577db0499d880171f324bdf43f9e19d...   2020-01-01  
1233963  33e9ec4b9da2563577db0499d880171f324bdf43f9e19d...   2020-01-01  
1234741  33e9ec4b9da2563577db0499d880171f324bdf43f9e19d...   2020-01-01  
...                                                    ...          ...  
1732750  33e9ec4b9da2563577db0499d880171f324bdf43f9e19d...   2020-01-01  
1733509  33e9ec4b9da2563577db0499d880171f324bdf43f9e19d...   2020-01-01  
1734227  33e9ec4b9da2563577db0499d880171f324bdf43f9e19d...   2020-01-01  
1734365  33e9ec4b9da2563577db0499d880171f324bdf43f9e19d...   2020-01-01  
1734448  33e9ec4b9da2563577db0499d880171f324bdf43f9e19d...   2020-01-01  

[656 rows x 8 columns]

In [35]:
cpf_total[cpf_total.CPF=="44198631883"]

CPF        nu_cpf
12128675  44198631883  4.419863e+10

In [37]:
csup[csup.nu_cpf==44198631883]

Empty DataFrame
Columns: [co_aluno, nu_cpf, sexo, cor_raca, cor_raca_ces, cor_raca_enem, conclusao_em, ano_conclusao_em, idade, ano_nascimento, in_enem, nota_cn, nota_ch, nota_lc, nota_mt, notas, nota_redacao, edupar, escpai, escmae, rdpc, log_rdpc, renda, qt_hab, imovel, trabalha, escola_ef, escola_em, co_curso, vinculo_2013, vinculo_2014, vinculo_2015, vinculo_2016, vinculo_2017, vinculo_2018, vinculo_2019, co_ies, categoria_adm, organizacao_acad, grau, modalidade, co_cine_rotulo, no_cine_rotulo, co_cine_area_detalhada, no_cine_area_detalhada, co_cine_area_especifica, no_cine_area_especifica, co_cine_area_geral, no_cine_area_geral, reserva_vagas_2013, turno_2013, turno_2014, turno_2015, turno_2016, turno_2017, turno_2018, turno_2019, financiamento_2013, financiamento_2014, financiamento_2015, financiamento_2016, financiamento_2017, financiamento_2018, financiamento_2019, fies_2013, fies_2014, fies_2015, fies_2016, fies_2017, fies_2018, fies_2019, prouni_2013, prouni_2014, prouni_2015, prouni_2016, prouni_2017, prouni_2018, prouni_2019, apoio_social_2013, apoio_social_2014, apoio_social_2015, apoio_social_2016, apoio_social_2017, apoio_social_2018, apoio_social_2019]
Index: []

[0 rows x 85 columns]

In [39]:
len(pd.unique(csup.nu_cpf))

2763108

In [42]:
len(pd.unique(cpf_total.nu_cpf))

54999886